Add checks for IPv4 and IPv6

vishalmohanty opened the pull request:
This diff does the following

- Adds checks to make sure `A` returns `IPv4` and `AAAA` returns `IPv6`.
- Adds unit tests for the IP checks
- Adds some additional unit tests for loops, empty non-terminal, non-existent domains

vishalmohanty committed:
Add checks for IPv4 and IPv6

zakird commented:
@spencerdrak can you take a first pass review?
@vishalmohanty can you please fix conflicts?

codyprime commented on modules/alookup/alookup.go:
```diff
@@ -43,6 +44,18 @@ func (s *Lookup) DoLookup(name, nameServer string) (interface{}, zdns.Trace, zdn
+// Verify that A record is indeed IPv4 and AAAA is IPv6
+func verifyAddress(ansType string, ip string) bool {
+	isIpv4 := net.ParseIP(ip).To4() != nil
+	isIpv6 := net.ParseIP(ip).To16() != nil
```
`To16()` doesn't make any guarantees the address is IPv6; it will return non-nil for both IPv4 and IPv6, it merely converts either IPv4 or IPv6 into 16-byte representation:

https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/net/ip.go;l=226

if `net.ParseIP(ip).To16()` returns non-nil, you can then check for `':'` in `ip` (the string) to verify it is IPv6

vishalmohanty replied:
Done. Thanks!

spencerdrak commented on modules/alookup/alookup.go:
```diff
@@ -68,11 +81,16 @@ func (s *Lookup) doLookupProtocol(name, nameServer string, dnsType uint16, candi
-					candidateSet[lowerCaseName] = append(candidateSet[lowerCaseName], ans)
-				} else if ok && dns.TypeCNAME == ansType {
-					cnameSet[lowerCaseName] = append(cnameSet[lowerCaseName], ans)
+				// Verify that the answer type matches requested type
```
Only thing I might bring up is that we've got some duplication going on here with what happens in each loop, it may be cleaner to pull that logic out. Thoughts?

gakiwate replied:
Is the answer always an A/AAAA/CNAME record? What if you query for a NS record? Does that go through a different logic?

vishalmohanty replied:
@spencerdrak moved it to `populateResults`.

@gakiwate `verifyAddress` only checks for A record containing IPv4 and AAAA containing IPv6. For any other record, we return true

spencerdrak reviewed (commented):
Overall looks good except for the two comments that are here, lmk what you think @vishalmohanty!

gakiwate commented on modules/alookup/alookup_test.go:
```diff
@@ -263,6 +263,94 @@ func TestDoLookup(t *testing.T) {
+		t.Errorf("Expected no results, got %v", res)
+	}
+
+	// Case 12: Loops in CNAME should terminate with no results
```
I wonder if this should terminate with no results or throw an error? For example, if we want to find how many CNAMEs chains have loops in them? Then this will hide those errors from us no? @spencerdrak any thoughts?

[Relevant RFC 1034](https://www.rfc-editor.org/rfc/rfc1034.html#section-3.6.2)
> CNAME chains should be followed and CNAME loops signalled as an error

gakiwate replied:
I would recommend checking how Google and Cloudflare deal with the loops and see if we are consistent with them?

vishalmohanty replied:
Returning `STATUS_ERROR` in case of loops now.

gakiwate commented on modules/alookup/alookup_test.go:
```diff
@@ -263,6 +263,94 @@ func TestDoLookup(t *testing.T) {
+
+	// Case 14: Non-existent domain
+	res, _, status, _ = l.DoLookup("nonexistent.example.com", "")
+	if status != zdns.STATUS_NO_ANSWER {
```
The error here should be NXDOMAIN no?

vishalmohanty replied:
Yes. Updated

gakiwate commented on modules/alookup/alookup_test.go:
```diff
@@ -263,6 +263,94 @@ func TestDoLookup(t *testing.T) {
+		t.Errorf("Expected no results, got %v", res)
+	}
+
+	// Case 13: Empty non-terminal domain should give no answer
```
This should be no answer but also no error? Does STATUS_NO_ANSWER imply NOERROR?

vishalmohanty replied:
Removed `STATUS_NO_ANSWER` altogether because it is not a standard dns error code. In all cases where we were returning `STATUS_NO_ANSWER`, we are now returning `NOERROR`

gakiwate reviewed (commented):
At a high level, my feedback is to make sure that the RCODEs (NXDOMAIN, NOERROR) that your test suite returns is consistent with how other resolvers do it/RFCs recommend to do it.

vishalmohanty commented:
Changes made:

- Removed `STATUS_NO_ANSWER` because it is not a standard dns error code. Returning `STATUS_NOERROR` in its place.
- If IPv4 or IPv6 lookup fail with a certain error (like `SERVFAIL`), we return the same error code.
- Added UTs for verifying we return the same error code as returned by miekgdns. The idea is to make all modules consistent with the expected response from an actual dns resolver.
- Fixed few lints.

spencerdrak reviewed (commented):
I think with all the suggested changes made, things look good. I'll hold until others have the chance to also review again.

gakiwate commented on modules/alookup/alookup_test.go:
```diff
@@ -23,36 +23,41 @@ import (
 	l, err := rlf.MakeLookup()
 	if l == nil || err != nil {
-		t.Error("Failed to initialize lookup")
+		panic("Failed to initialize lookup")
```
@vishalmohanty I am assuming this change is to make it consistent with other lookup files?

vishalmohanty replied:
Thanks for catching this. I'm doing `t.Error` now.

gakiwate commented on modules/alookup/alookup_test.go:
```diff
@@ -257,9 +349,140 @@ func TestDoLookup(t *testing.T) {
+
+	// Verify empty non-terminal returns no answer
+	res, _, status, _ := l.DoLookup("intermediate.example.com", "")
+	if status != zdns.STATUS_NOERROR {
```
You also need to check that no answers were returned. The NOERROR does not mean no answers were returned.

vishalmohanty replied:
That is done in the next line.

gakiwate commented on modules/alookup/alookup_test.go:
```diff
@@ -257,9 +349,140 @@ func TestDoLookup(t *testing.T) {
+	glf.IPv4Lookup = true
+	res, _, status, _ := l.DoLookup("nonexistent.example.com", "")
+	if status != zdns.STATUS_NXDOMAIN {
+		t.Errorf("Expected STATUS_NOERROR status, got %v", status)
```
typo?

vishalmohanty replied:
Yes, Thanks for the catch!

codyprime commented on modules/alookup/alookup.go:
```diff
@@ -43,13 +44,47 @@ func (s *Lookup) DoLookup(name, nameServer string) (interface{}, zdns.Trace, zdn
+// Verify that A record is indeed IPv4 and AAAA is IPv6
+func verifyAddress(ansType string, ip string) bool {
+	isIpv4 := net.ParseIP(ip).To4() != nil
+	isIpv6 := net.ParseIP(ip).To16() != nil && strings.Contains(ip, ":")
```
One more comment, if you want a minor optimization.  Since:
* You are always calling `To16()`
* `To16()` will guarantee, if non-nil, that `ip` is a valid IP

You don't need the call to `To4()`.  Logic like this would suffice and save a conversion call:

```go
var isIpv4, isIpv6 bool
if net.ParseIP(ip).To16() != nil {
    isIpv6 = strings.Contains(ip, ":")
    isIpv4 = !isIpv6
}
```

But even beyond that, I don't think the call to `To16()` is even needed; `ParseIP` will return `nil` if is not a valid textual representation of an IP address [1].  So I think this can get abbreviated down to:

```go
var isIpv4, isIpv6 bool
if net.ParseIP(ip) != nil {
    isIpv6 = strings.Contains(ip, ":")
    isIpv4 = !isIpv6
}
```

[1] https://pkg.go.dev/net#ParseIP

vishalmohanty replied:
Done. Also initialised isIpv4 and isIpv6 to false.

vishalmohanty committed:
Address comments

vishalmohanty committed:
Initialize bools to false
